File: Applied Course/4.Problem Solving/4.Problems on Stacks and Queues/7.Reverse a stack using recursion.c
/*Write a program to reverse a stack using recursion.and you can only use the following ADT functions on Stack S:
isEmpty(S)
push(S)
pop(S)*/

#include <stdio.h>
int stack[100];
int top=-1;

void push(int x)
{
	stack[++top]=x;
}
int pop()
{
	if(top==-1)
		return 0;
	int d=stack[top--];
	return d;
}
int isEmpty()
{
	if(top==-1)
		return 0;
	else
		return 1;
}
int peek()
{
	if(top!=-1)
		return stack[top];
}

int size()
{
	int count=0;
	int top1=top;
	while((top1--)>-1)
		count++;
	return count;
}

void insert(int x)
{
	if(size()==0){
		push(x);
	}

	else{
		char c=pop();
		insert(x);

		push(c);
	}
}

void reverse()
{
	if(size()>0)
	{
		char c=pop();
		reverse();

		insert(c);
	}
}
int main()
{
	push(1);
	push(2);
	push(3);
	push(4);
	int n=size();
	printf("Before reverse\n");
	for (int i = 0; i < n; ++i)
	{
		printf("%d\n",pop());
	}
	top=n-1;
	reverse();
	n=size();
	printf("After reverse\n");
	for(int i=0;i<n;i++)
		printf("%d\n",pop());
	return 0;
}

File: Applied Course/4.Problem Solving/4.Problems on Stacks and Queues/7.Reverse a stack using recursion.py
import sys
class node: 

    def __init__(self, info): 
        self.info = info  
        self.next = None 


class Stack: 

    def __init__(self): 
        self.top = None
        
    
    def isEmpty(self):
        if self.top is None:
            return True
        return False
    
    def push(self,data):
        self.temp=node(data)
        if self.temp is None:
            print("Stack overflow")
            return
        self.temp.next=self.top
        self.top=self.temp
    
    def pop(self):
        if self.isEmpty():
            print("Stack Underflow")
            sys.exit(0)
        d=self.top.info
        self.top=self.top.next    
        return d
    
    def peek(self):
        if self.isEmpty():
            print("Stack Underflow")
            sys.exit(0)
        d=self.top.info
        return d
    def display(self):
        if self.isEmpty():
            print("Stack Underflow")
            sys.exit(0)
        self.p=self.top
        while self.p is not None:
            print(self.p.info)
            self.p=self.p.next
def insert(s,c):
    if s.isEmpty() is True:
        s.push(c)
    else:
        d=s.pop()
        insert(s,c)
        s.push(d)
    return s
def reverse(s):
    if s.isEmpty() is not True:
        c=s.pop()
        reverse(s)
        insert(s,c)
        return s
if __name__=='__main__':
    stack=Stack()
    stack.push(1)
    stack.push(2)
    stack.push(3)
    stack.push(4)
    stack.push(10)
    print("Before reversing")
    stack.display()
    stack=reverse(stack)
    print("After reversing")
    stack.display()
    
